presidio_evaluator/evaluation/span_evaluator.py

```diff
@@ -682,3 +604,162 @@ def _calculate_metrics(
         recall = self.recall(tp=true_positives, num_annotated=num_annotated)
         f_beta = self.f_beta(precision=precision, recall=recall, beta=beta)
         return precision, recall, f_beta
+
+
+    def _match_predictions_with_annotations(
+        self,
+        annotation_spans: List[Span],
+        prediction_spans: List[Span],
+        evaluation_result: EvaluationResult,
+        per_type: bool = True,
+    ) -> EvaluationResult:
+        """
+        Match predictions to annotations in a single pass.
+        For each annotation, incrementally add prediction spans as long as they improve the IoU.
+        Only tracks per-entity type metrics.
+
+        :param annotation_spans: List of annotation Span objects
+        :param prediction_spans: List of prediction Span objects
+        :param evaluation_result: EvaluationResult object to update
+        :return: Updated EvaluationResult
+        """
+        matched_predictions = set()
+
+        if not evaluation_result.model_errors:
+            evaluation_result.model_errors = []
+
+        # Process each annotation span
+        for ann_span in annotation_spans:
```

Comment:
We discussed counting num_actual and num_predicted based on the actual number of spans that we used. Currently, it still counts them a priori. Let's discuss if you found that this could be an issue.